### PR TITLE
Self Update

### PR DESCRIPTION
**Pull request by jacderida**
Hi Spandan/Stephen,

This uses the [self_update](https://crates.io/crates/self_update) crate to pull in new releases from GitHub.

On my fork there are a couple of [releases](https://github.com/jacderida/safe_vault/releases). I've reverted the version change commits I used to produce these releases.

See the message on [this commit](https://github.com/maidsafe/safe_vault/pull/788/commits/1f73aa90e242265923807fc519f3458f616a1ccb) for the detail on how the updating process works. It's as we proposed on Slack.

### Before Update

At this point in time, only version 0.18.1 was available on GitHub.

Here is the output from running version 0.18.1:
```
vagrant@stretch:~$ export RUST_LOG=debug
vagrant@stretch:~$ ./safe_vault --version
<other output snipped>
safe-vault 0.18.1
vagrant@stretch:~$ ./safe_vault 
[2019-08-12T14:06:55Z INFO  safe_vault::detail] Checking for updates...
[2019-08-12T14:06:56Z DEBUG safe_vault::detail] Target for update is x86_64-unknown-linux-musl
[2019-08-12T14:06:56Z DEBUG safe_vault::detail] Found releases: [
    Release {
        name: "safe-vault",
        body: "Command line interface for the SAFE Network",
        tag: "0.18.1",
        date_created: "2019-07-24T12:18:44Z",
        assets: [
            ReleaseAsset {
                download_url: "https://github.com/jacderida/safe_vault/releases/download/0.18.1/safe_vault-0.18.1-x86_64-apple-darwin.tar",
                name: "safe_vault-0.18.1-x86_64-apple-darwin.tar",
            },
            ReleaseAsset {
                download_url: "https://github.com/jacderida/safe_vault/releases/download/0.18.1/safe_vault-0.18.1-x86_64-pc-windows-gnu.tar",
                name: "safe_vault-0.18.1-x86_64-pc-windows-gnu.tar",
            },
            ReleaseAsset {
                download_url: "https://github.com/jacderida/safe_vault/releases/download/0.18.1/safe_vault-0.18.1-x86_64-unknown-linux-musl.tar",
                name: "safe_vault-0.18.1-x86_64-unknown-linux-musl.tar",
            },
        ],
    },
]

Checking target-arch... x86_64-unknown-linux-musl
Checking current version... v0.18.1
Checking latest released version...
v0.18.1
Update status: `0.18.1`!
[2019-08-12T14:06:56Z INFO  safe_vault::detail]
```

### After Update

Version 0.18.2 was then uploaded to GitHub.

File size before the update:
```
vagrant@stretch:~$ ls -al
total 18036
<other output snipped>
-rwxr-xr-x 1 vagrant vagrant 18438680 Aug 12 14:02 safe_vault
```

Run version 0.18.1 again:
```
vagrant@stretch:~$ ./safe_vault 
[2019-08-12T14:34:02Z INFO  safe_vault::detail] Checking for updates...
Checking target-arch... x86_64-unknown-linux-musl
Checking current version... v0.18.1
Checking latest released version... v0.18.2
New release found! v0.18.1 --> v0.18.2
New release is compatible

safe_vault release status:
  * Current exe: "/home/vagrant/safe_vault"
  * New exe release: "safe_vault-0.18.2-x86_64-unknown-linux-musl.tar"
  * New exe download url: "https://github.com/jacderida/safe_vault/releases/download/0.18.2/safe_vault-0.18.2-x86_64-unknown-linux-musl.tar"

The new release will be downloaded/extracted and the existing binary will be replaced.
Downloading...
17.59 MB / 17.59 MB [==============================================================] 100.00 % 4.80 MB/s   ... Done
Extracting archive... Done
Replacing binary file... Done
Update status: `0.18.2`!
Vault has been updated. Please restart.
```

File size after update:
```
vagrant@stretch:~$ ls -al
total 18036
<other output snipped>
-rwxr-xr-x 1 vagrant vagrant 18438688 Aug 12 14:28 safe_vault
```

Run again on version 0.18.2:
```
vagrant@stretch:~$ ./safe_vault 
[2019-08-12T14:34:23Z INFO  safe_vault::detail] Checking for updates...
Checking target-arch... x86_64-unknown-linux-musl
Checking current version... v0.18.2
Checking latest released version... v0.18.2
Update status: `0.18.2`!
[2019-08-12T14:34:24Z INFO  safe_vault::detail]
```

## QA

I've only tested this on Linux. Would be good to have some QA on other platforms.

### Instructions

* Download and untar version 0.0.1 for [macOS](https://github.com/jacderida/safe_vault/releases/download/0.18.1/safe_vault-0.18.1-x86_64-apple-darwin.tar) and [Windows](https://github.com/jacderida/safe_vault/releases/download/0.18.1/safe_vault-0.18.1-x86_64-pc-windows-gnu.tar) on each target platform respectively
* Enable some additional logging output with `export RUST_LOG=info` or `set RUST_LOG=info` on Windows
* Confirm you're running on 0.18.1 with `./safe_vault --version`
* Get the size of the binary using `ls` or `dir` on Windows
* Run the update with `./safe_vault`
* After the update run `./safe_vault --version` to confirm you're on version 0.0.2
* Get the size of the updated binary using `ls` or `dir` on Windows
* Run the command again with `./safe_vault`

### Acceptance Criteria

* Version 0.18.1 of `./safe_vault` should report the correct version.
* When running `./safe_vault` there should be some output to indicate the availability of a new version.
* The user should *not* be prompted to confirm accepting the update.
* The user should be informed that they need to restart after the update.
* After updating, the `safe_vault` binary should report the correct version.

Cheers,

Chris

### CODE REVIEW

File: src/bin/safe_vault.rs [hunk 1]
```diff
@@ -58,6 +58,9 @@
     variant_size_differences
 )]
 
+#[macro_use]
```
Comment (ustulation) [inline src/bin/safe_vault.rs]: is `macro_use` necessary ? You don't seem to be using any macros from that crate ? Either that or I probably missed locating one :)

Comment (jacderida) [reply src/bin/safe_vault.rs]: `log::info!("Current version is {}", cargo_crate_version!());`

It's for this.

Comment (ustulation) [reply src/bin/safe_vault.rs]: Ah macro inside a macro ! No wonder i missed - thanks, OK now.

**Comment by ustulation** [comment]
I've approved it now but do change
> After updating, the safe_vault binary should have a different size.

as discussed on slack.
@jacderida

**Comment by S-Coyle** [comment]
Note that running `./safe_vault --version` will actually trigger the update. This isn't a bug though - this is expected behaviour.

- [x] MacOS - acceptance criteria met
- [x] Windows - acceptance criteria met

**Comment by jacderida** [comment]
Thanks @ustulation

I've now removed that from the acceptance critiera.

**Comment by ustulation** [comment]
Cheers @jacderida ,
Once @S-Coyle is OK he can approve and merge it then.

**Pull request merged**